File: solutions/python/src/shared/outbox.py
"""Port (R8): the application service writes the aggregate AND its integration events through
this port inside one unit of work. Nothing here talks to a broker."""
from datetime import datetime, timezone
from typing import Protocol

from shared.integration_event import IntegrationEvent


class Outbox(Protocol):
    def add(self, event: IntegrationEvent) -> None: ...
    def pending(self) -> list[IntegrationEvent]: ...
    def mark_published(self, event_id: str) -> None: ...


class InMemoryOutbox:
    def __init__(self) -> None:
        self._rows: dict[str, tuple[IntegrationEvent, str | None]] = {}

    def add(self, event: IntegrationEvent) -> None:
        self._rows[event.event_id] = (event, None)

    def pending(self) -> list[IntegrationEvent]:
        return [event for event, published_at in self._rows.values() if published_at is None]

    def mark_published(self, event_id: str) -> None:
        if event_id in self._rows:
            event, _ = self._rows[event_id]
            self._rows[event_id] = (event, datetime.now(timezone.utc).isoformat())

    def all(self) -> list[IntegrationEvent]:
        """Test helper: everything ever written, published or not."""
        return [event for event, _ in self._rows.values()]

File: solutions/node/src/shared/Outbox.ts
import type { IntegrationEvent } from './IntegrationEvent.ts';

// Port (R8). The application service writes the aggregate AND its integration events
// through this port inside one unit of work. Nothing here talks to a broker.
export interface Outbox {
  add(event: IntegrationEvent): Promise<void>;
  pending(): Promise<IntegrationEvent[]>;
  markPublished(eventId: string): Promise<void>;
}

export class InMemoryOutbox implements Outbox {
  private readonly rows = new Map<string, { event: IntegrationEvent; publishedAt?: string }>();

  async add(event: IntegrationEvent): Promise<void> {
    this.rows.set(event.eventId, { event });
  }

  async pending(): Promise<IntegrationEvent[]> {
    return [...this.rows.values()].filter((r) => !r.publishedAt).map((r) => r.event);
  }

  async markPublished(eventId: string): Promise<void> {
    const row = this.rows.get(eventId);
    if (row) row.publishedAt = new Date().toISOString();
  }

  // Test helper: everything ever written, published or not.
  all(): IntegrationEvent[] {
    return [...this.rows.values()].map((r) => r.event);
  }
}
